Source/ProgressionSystemRuntime/Private/Data/PSWorldSubsystem.cpp

```diff
@@ -37,3 +45,113 @@ void UPSWorldSubsystem::SetProgressionSystemComponent(UPSHUDComponent* MyProgres
+	{
+		ProgressionDataTableInternal = PSDataAssetInternal->GetProgressionDataTable();
+		LoadGameFromSave();	
+	} else
+	{
+		UE_LOG(LogProgressionSystem, Warning, TEXT("PSDataAssetInternal null"))
+	}
+}
+
+void UPSWorldSubsystem::OnCharacterPossessed(APawn* MyPawn)
+{
+	if (APlayerCharacter* MyPlayerCharacter = Cast<APlayerCharacter>(MyPawn))
+	{
+		MyPlayerCharacter->OnPlayerTypeChanged.AddUniqueDynamic(this, &ThisClass::OnPlayerTypeChanged);
+
+		//Unsubscribe to ignore null events call
+		AMyPlayerController* MyPC = UMyBlueprintFunctionLibrary::GetLocalPlayerController();
+		MyPC->GetOnNewPawnNotifier().RemoveAll(this);
+	}
+}
+
+void UPSWorldSubsystem::OnPlayerTypeChanged(FPlayerTag PlayerTag)
+{
+	if (SaveGameInstanceInternal)
+	{
+		for (const auto& KeyValue : SaveGameInstanceInternal->SavedProgressionRows)
+		{
+			FName Key = KeyValue.Key;
+			FPSRowData RowData = KeyValue.Value;
+
+			if (RowData.Map == UMyBlueprintFunctionLibrary::GetLevelType() && RowData.Character == PlayerTag)
+			{
+				SavedProgressionRowDataInternal = SaveGameInstanceInternal->SavedProgressionRows[Key];
+			}
+		}
+		SaveDataAsync();
+	}
+	OnCurrentRowDataChanged.Broadcast(SavedProgressionRowDataInternal);
+}
+
+void UPSWorldSubsystem::LoadGameFromSave()
+{
+	// Check if the save game file exists
+	if (UGameplayStatics::DoesSaveGameExist(SaveGameInstanceInternal->GetSaveSlotName(), SaveGameInstanceInternal->GetSaveSlotIndex()))
+	{
+		SaveGameInstanceInternal = Cast<UPSSaveGameData>(UGameplayStatics::LoadGameFromSlot(SaveGameInstanceInternal->GetSaveSlotName(), SaveGameInstanceInternal->GetSaveSlotIndex()));
+	}
+	else
+	{
+		// Save file does not exist
+		// do initial load from data table
+		TMap<FName, FPSRowData> SavedProgressionRows;
+		UMyDataTable::GetRows(*ProgressionDataTableInternal, SavedProgressionRows);
+		SaveGameInstanceInternal = Cast<UPSSaveGameData>(UGameplayStatics::CreateSaveGameObject(UPSSaveGameData::StaticClass()));
+
+		if (SaveGameInstanceInternal)
+		{
+			SaveGameInstanceInternal->SavedProgressionRows = SavedProgressionRows;
+		}
+	}
+	SetFirstElemetAsCurrent();
+}
+
+void UPSWorldSubsystem::SetFirstElemetAsCurrent()
+{
+	if (SaveGameInstanceInternal)
+	{
+		// Create an iterator for the TMap
+		TMap<FName, FPSRowData>::TIterator Iterator = SaveGameInstanceInternal->SavedProgressionRows.CreateIterator();
+
+		// Check if the iterator is pointing to a valid position
+		if (Iterator)
+		{
+			// Access the key and value of the first element
+			SavedProgressionRowDataInternal = Iterator.Value();
+			SavedProgressionRowDataInternal.IsLevelLocked = false;
+			Iterator.Value().IsLevelLocked = false;
+		}		
+		else
+		{
+			// The TMap is empty
+			UE_LOG(LogTemp, Warning, TEXT("The TMap is empty."));
+		}		
+	}
+	SaveDataAsync();
+}
+
+void UPSWorldSubsystem::SaveDataAsync()
+{
+	UGameplayStatics::AsyncSaveGameToSlot(SaveGameInstanceInternal, SaveGameInstanceInternal->GetSaveSlotName(), SaveGameInstanceInternal->GetSaveSlotIndex());
```

Comment:
`SaveGameInstanceInternal` is not checked on null. If it 100% has to be valid, then add above `checf(SaveGameInstanceInternal, TEXT("ERROR: Text")`, so it will crash with your message in log instead of silent crash.
If there is small chance it can be null because of wrong setup in blueprints, then instead of crashing, prefer if (ensureMsgf(SaveGameInstanceInternal, TEXT("ASSERT: Text")`.
But never leave pointers unhandled as we have there

Reply:
done

Reply:
done, if you will any missing spots - let me know